cmd/gen-release-notes/main.go

```diff
@@ -110,19 +98,171 @@ func main() {
+		"prevMilestone":               prevMilestone,
+		"changeLogSince":              changeLogSince,
+		"content":                     content,
+		"k8sVersion":                  k8sVersion,
+		"changeLogVersion":            markdownVersion,
+		"majorMinor":                  majorMinor,
+		"EtcdVersion":                 getGoModVersion("etcd", repo, milestone),
+		"ContainerdVersion":           getGoModVersion("containerd", repo, milestone),
+		"RuncVersion":                 getGoModVersion("runc", repo, milestone),
+		"CNIPluginsVersion":           getImageTagVersion("cni-plugins", repo, milestone),
+		"MetricsServerVersion":        getImageTagVersion("metrics-server", repo, milestone),
+		"TraefikVersion":              getImageTagVersion("traefik", repo, milestone),
+		"CoreDNSVersion":              getImageTagVersion("coredns", repo, milestone),
+		"IngressNginxVersion":         getDockerfileVersion("rke2-ingress-nginx", repo, milestone),
+		"HelmControllerVersion":       getGoModVersion("helm-controller", repo, prevMilestone),
+		"FlannelVersionRKE2":          getImageTagVersion("flannel", repo, milestone),
+		"FlannelVersionK3S":           getGoModVersion("flannel", repo, milestone),
+		"CalicoVersion":               calicoVersion,
+		"CalicoVersionTrimmed":        calicoVersionTrimmed,
+		"CiliumVersion":               getImageTagVersion("cilium-cilium", repo, milestone),
+		"MultusVersion":               getImageTagVersion("multus-cni", repo, milestone),
+		"KineVersion":                 getGoModVersion("kine", repo, milestone),
+		"SQLiteVersion":               sqliteVersionBinding,
+		"SQLiteVersionReplaced":       strings.ReplaceAll(sqliteVersionBinding, ".", "_"),
+		"LocalPathProvisionerVersion": getImageTagVersion("local-path-provisioner", repo, milestone),
 	}); err != nil {
-		fmt.Println(err)
-		os.Exit(1)
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	flag.Usage = func() {
+		w := os.Stderr
+		for _, arg := range os.Args {
+			if arg == "-h" {
+				w = os.Stdout
+				break
+			}
+		}
+		fmt.Fprintf(w, usage, version, name)
+	}
+
+	flag.BoolVar(&vers, "v", false, "")
+	flag.BoolVar(&debug, "d", false, "")
+	flag.StringVar(&ghToken, "t", "", "")
+	flag.StringVar(&repo, "r", "", "")
+	flag.StringVar(&milestone, "m", "", "")
+	flag.StringVar(&prevMilestone, "p", "", "")
+	flag.Parse()
+
+	if err := run(); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func getGoModVersion(libraryName, repo, branchVersion string) string {
+	repoName := "k3s-io/k3s"
+	if repo == "rke2" {
+		repoName = "rancher/rke2"
+	}
+	goModURL := "https://raw.githubusercontent.com/" + repoName + "/" + branchVersion + "/go.mod"
+	submatch := findInURL(goModURL, "", libraryName)
+	logrus.Info(submatch)
+	for _, line := range submatch {
+		trimmedLine := strings.TrimSpace(line)
+		// use replace section if found
+		if strings.Contains(trimmedLine, "=>") {
+			libVersionLine := strings.Split(trimmedLine, " ")
+			return libVersionLine[3]
```

Comment:
I think this will return the wrong thing if there is a versioned replace:
```go.mod
replace (
    example.com/bad/thing v1.4.5 => example.com/good/thing v1.4.5
)
```

Although you could just check how many parts there are in the split line, I'd encourage you to use the go modfile library for this parsing instead, if it's not too much work to drop in.

Reply:
I will give the go modfile library a shot, to avoid all confusion, thanks!